File: internal/rules/no_empty_pattern/no_empty_pattern_test.go
package no_empty_pattern

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/typescript/rules/fixtures"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoEmptyPatternRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&NoEmptyPatternRule,
		[]rule_tester.ValidTestCase{
			{Code: `var {a} = obj;`},
			{Code: `var [a] = arr;`},
			{Code: `var {a = 1} = obj;`},
			{Code: `var [a = 1] = arr;`},
			{Code: `function foo({a}) {}`},
			{Code: `function foo([a]) {}`},
			{Code: `var {a: {b}} = obj;`},
			{Code: `var {a: [b]} = obj;`},
			// allowObjectPatternsAsParameters: direct parameter
			{
				Code:    `function foo({}) {}`,
				Options: map[string]interface{}{"allowObjectPatternsAsParameters": true},
			},
			// allowObjectPatternsAsParameters: parameter with empty object default
			{
				Code:    `function foo({} = {}) {}`,
				Options: map[string]interface{}{"allowObjectPatternsAsParameters": true},
			},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: `var {} = obj;`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 5},
				},
			},
			{
				Code: `var [] = arr;`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 5},
				},
			},
			{
				Code: `var {a: {}} = obj;`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 9},
				},
			},
			{
				Code: `var {a: []} = obj;`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 9},
				},
			},
			{
				Code: `function foo({}) {}`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 14},
				},
			},
			{
				Code: `function foo([]) {}`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 14},
				},
			},
			// allowObjectPatternsAsParameters: non-empty default should still report
			{
				Code:    `function foo({} = {a: 1}) {}`,
				Options: map[string]interface{}{"allowObjectPatternsAsParameters": true},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 14},
				},
			},
			// allowObjectPatternsAsParameters: non-object default should still report
			{
				Code:    `function foo({} = bar) {}`,
				Options: map[string]interface{}{"allowObjectPatternsAsParameters": true},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 14},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint/rules/no-empty-pattern.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-empty-pattern', {
  valid: [
    'var {a} = obj;',
    'var [a] = arr;',
    'var {a = 1} = obj;',
    'var [a = 1] = arr;',
    'function foo({a}) {}',
    'function foo([a]) {}',
    'var {a: {b}} = obj;',
    'var {a: [b]} = obj;',
    // allowObjectPatternsAsParameters: direct parameter
    {
      code: 'function foo({}) {}',
      options: { allowObjectPatternsAsParameters: true },
    },
    // allowObjectPatternsAsParameters: parameter with empty object default
    {
      code: 'function foo({} = {}) {}',
      options: { allowObjectPatternsAsParameters: true },
    },
  ],
  invalid: [
    {
      code: 'var {} = obj;',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'var [] = arr;',
      errors: [{ messageId: 'unexpected' }],
    },
    // Nested empty patterns
    {
      code: 'var {a: {}} = obj;',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'var {a: []} = obj;',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'function foo({}) {}',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'function foo([]) {}',
      errors: [{ messageId: 'unexpected' }],
    },
    // allowObjectPatternsAsParameters: non-empty default should still report
    {
      code: 'function foo({} = {a: 1}) {}',
      options: { allowObjectPatternsAsParameters: true },
      errors: [{ messageId: 'unexpected' }],
    },
    // allowObjectPatternsAsParameters: non-object default should still report
    {
      code: 'function foo({} = bar) {}',
      options: { allowObjectPatternsAsParameters: true },
      errors: [{ messageId: 'unexpected' }],
    },
  ],
});
